### DISCUSSION TITLE
Licensing files

### DISCUSSION THREAD

**Discussion by LeonardoBeleffi** [community/community#48493] [Other Feature Feedback, Questions, & Ideas] [answered]
### Good morning, after not finding an answer I decided to post my question here. How can I proceed if I want to put a default license in my repo, meaning that I want it to be applied to all the files except the ones that specifies a different license inside (in the header)?

**Comment by Kaparrs** [comment]
- You can add a default license when you're creation a repository 
- If you want to add a license to an existing repository, you can find more info here: [Github Docs - Adding a license to a repository](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository)

**Reply by LeonardoBeleffi** [reply]
Thanks!
So, if a file specifies another license it overrides the default one regardless which one is the more restrictive?

**Reply by Kaparrs** [reply]
I'm not really sure about that. You can just specify it at the top of the file, because to my knowledge GitHub doesn't have a system for licensing files separately as that isn't a common practice.

**Reply by LeonardoBeleffi** [reply]
Yes yes, I ment the file specified it inside at the top, before anything else.

**Comment by Huseyin-Cinar** [comment]
**Hey @LeonardoBeleffi**
- You have one chance when you're selecting license for a repository.
- This license including all repository.
- You can not selecting different licenses for repository's documents.
- **If you confirm my answer i will so happy. That's all, have a nice day :)**

**Reply by LeonardoBeleffi** [reply]
So if I have to put different files in my repo, and not all files have the same license what do I do?

**Reply by Huseyin-Cinar** [reply]
- I think this problem's only soltuion different files putting different repositories.

**Reply by LeonardoBeleffi** [reply]
They belong to the same one tho...

**Reply by Huseyin-Cinar** [reply]
**If you confirm my answer i will so happy. That's all, have a nice day :)**

**Reply by Huseyin-Cinar** [reply]
> They belong to the same one tho...

- That doesn't change anything.

**Reply by Kaparrs** [reply]
> They belong to the same one tho...

You could put them is submodules

**Reply by LeonardoBeleffi** [reply]
> > They belong to the same one tho...
> 
> - That doesn't change anything.

Yes it does, making like 4 repos for files concerning the same project is horrific...

**Reply by LeonardoBeleffi** [reply]
> > They belong to the same one tho...
> 
> You could put them is submodules

Thought about that but, as I said, it's the same project... I feel like that would be a very unhappy solution...

**Comment by SmashedFrenzy16** [comment]
To create a default license, just check a box for creating a license when creating a repo.

**Reply by LeonardoBeleffi** [reply]
And if I put two files in the repo, and inside one of the two its license (different from the "repo's one") will the two files obey the same license (the repo's one) or will they obey to the two respective licenses (the one with the licence inside will obey the licence inside, and the other one will obey the repo's one)?

**Reply by SmashedFrenzy16** [reply]
I think they should obey their respective ones then, as I don't think it can be both.

**Reply by LeonardoBeleffi** [reply]
I hope you're right, thanks!

**Reply by SmashedFrenzy16** [reply]
No problem! I am always happy to help!

**Comment by airtower-luna** [comment]
You can have files with different licenses in one repository (it's not that rare), but it's important to make clear which is which.

If you have a `LICENSE` file with one license at the top level and then a license header that says otherwise it'll be confusing. I'd stay clear of using such a repository (and thus contributing), unless I have money and motivation to get a legal consultation first, or the owner clarifies the license.

How to best describe licenses is a widely discussed question. [SPDX](https://spdx.dev/) identifiers are meant to make license information unambiguous and machine-readable, the [REUSE project](https://reuse.software/) offers tools to make sure all files contain clear information.

I'd say the minimum is:
* Make sure each source file has an appropriate license header.
* If you have code under multiple licenses in your repository, don't put *just one* of them into the `LICENSE` file, because that'll likely cause confusion.

I've seen different descriptions for multiple licenses in practice, both `LICENSE` files containing multiple licenses, and ones listing which files are under which license (which could include an "all other files" section) and referring to other files for the actual license text.

**Reply by sdruskat** [reply]
See also #15033 as related.

**Comment by Satyam6024** [comment]
You can apply a default license to your repository by placing a LICENSE file in the root. This license will apply to all files unless they explicitly state otherwise in their file headers.

In your README.md, clarify:
```
This repository is licensed under the MIT License by default. 
Specific files may state a different license in their headers, which takes precedence.
```

For exceptions, add a license comment at the top of the file (e.g., // SPDX-License-Identifier: Apache-2.0). This ensures clarity for users and legal tools.

**Comment by viancodes** [comment, marked as answer]
Good question! Here's how you can handle default licensing in your GitHub repository:

✅ 1. **Add a LICENSE file in the root of your repo**
- This license will apply to all files in your project **by default**, unless a file or folder explicitly states something different.
- GitHub automatically recognizes this as the main license for your repo.

✅ 2. **Override in specific files (optional)**
- If you want certain files to have a different license, add a license notice at the top of those files like:
```
// SPDX-License-Identifier: MIT
```
or
```
// Licensed under Apache-2.0 instead of the project’s default license.
```

✅ 3. **Use SPDX Identifiers**
- They’re short, standard tags that tools and humans can easily understand.
- More info: https://spdx.org/licenses/

📝 In summary:
- The LICENSE file applies by default.
- You can override it per-file using headers or SPDX tags.
- This approach is widely used in open source projects.

Let me know if you want help picking the right license for your use case!
